libs/sdk-ui-pivot/src/impl/data/rowFactory.ts

```diff
@@ -0,0 +1,229 @@
+): IGridRow {
+    const row: IGridRow = {
+        headerItemMap: {},
+    };
+
+    tableDescriptor.headers.sliceCols.forEach((rowHeader, rowHeaderIndex) => {
+        const { isSubtotal, field, value, rowHeaderDataItem } = getCell(
+            rowHeaderData,
+            rowIndex,
+            rowHeader,
+            rowHeaderIndex,
+            intl,
+        )!;
+        if (isSubtotal) {
+            row.type = ROW_SUBTOTAL;
+
+            if (!row.subtotalStyle) {
+                row.subtotalStyle = subtotalStyles[rowHeaderIndex] ?? undefined;
+            }
+        }
+
+        row[field] = value;
+        row.headerItemMap[field] = rowHeaderDataItem as IMappingHeader;
+    });
+
+    if (!tableDescriptor.hasDataLeafCols()) {
+        // table has no leaf columns - it is a row-only table listing a bunch of
+        // attribute element values. no point in going further as there is no additional
+        // data to show
+        return row;
+    }
+
+    cellData.forEach((cell: DataValue, cellIndex: number) => {
+        const colId = tableDescriptor.headers.leafDataCols[cellIndex].id;
+
+        invariant(colId);
+
+        row[colId] = cell;
+    });
+
+    return row;
+}
+
+export function getRowTotals(
+    tableDescriptor: TableDescriptor,
+    dv: DataViewFacade,
+    intl: IntlShape,
+): IGridTotalsRow[] | null {
+    if (!dv.rawData().hasTotals()) {
+        return null;
+    }
+
+    const totals = dv.rawData().totals();
+
+    invariant(totals && totals.length > 0);
+
+    const colGrandTotals = totals[0];
+    const colGrandTotalDefs = dv.definition.dimensions[0].totals;
+
+    invariant(colGrandTotalDefs);
+
+    const grandTotalColDescriptor = tableDescriptor.getGrandTotalCol();
+    const grandTotalAttrDescriptor = grandTotalColDescriptor.attributeDescriptor;
+    const leafColumns = tableDescriptor.zippedLeaves;
+
+    return colGrandTotals.map((totalsPerLeafColumn: string[], totalIdx: number) => {
+        const grandTotalName = grandTotalAttrDescriptor.attributeHeader.totalItems![totalIdx].totalHeaderItem
+            .name;
+        const measureCells: Record<string, string> = {};
+        const calculatedForColumns: string[] = [];
+        const calculatedForMeasures = colGrandTotalDefs
+            .filter((total) => {
+                return (
+                    grandTotalName === total.type &&
+                    total.attributeIdentifier === grandTotalAttrDescriptor.attributeHeader.localIdentifier
+                );
+            })
+            .map((total) => total.measureIdentifier);
+
+        totalsPerLeafColumn.forEach((value, idx) => {
+            const [leafDescriptor] = leafColumns[idx];
+
+            // if code bombs here then there must be something wrong in the table / datasource code because
+            // totals cannot (by definition) appear in a table that does not have measures - yet here we are
+            // processing totals
+            invariant(isDataColLeaf(leafDescriptor));
```

Comment:
Nitpick: a lot of invariants have a missing message (which in my opinion is a better development experience than reading the error description in the comment)

Reply:
You could argue thou, that those messages bloat the bundle (for cases that will hopefully not happen) while comments do not :-/

Reply:
This should be probably solvable at the bundler level? Well, it's not really important, I just like when errors are as much descriptive as possible, because they can point you in the right direction even without checking the code & comments 😅

Reply:
i normally do this too for invariants that are on public API boundary. in this case, it's all in the internals and these invariants are there to catch bugs in the table code itself. if any of the invariant bombs, it does not mean the SDK user provided bad input.. it's that there is a bug in the table code itself. whoever ends up debugging can find in the comments possible reason why the invariant failed - where roughly did things go wrong and resulted in the bad invalid invariant.

that was the thinking behind.. these invariants are essentially `assert`s